Basic isometry types

mweatherley opened the pull request:
# Objective

Introduce isometry types for describing relative and absolute position in mathematical contexts.

## Solution

For the time being, this is a very minimal implementation. This implements the following faculties for two- and three-dimensional isometry types:
- Identity transformations
- Creation from translations and/or rotations
- Inverses
- Multiplication (composition) of isometries with each other
- Application of isometries to points (as vectors)
- Conversion of isometries to affine transformations

There is obviously a lot more that could be added, so I erred on the side of adding things that I knew would be useful, with the idea of expanding this in the near future as needed.

(I also fixed some random doc problems in `bevy_math`.)

---

## Design

One point of interest here is the matter of if/when to use aligned types. In the implementation of 3d isometries, I used `Vec3A` rather than `Vec3` because it has no impact on size/alignment, but I'm still not sure about that decision (although it is easily changed). 

For 2d isometries — which are encoded by four floats — the idea of shoving them into a single 128-bit buffer (`__m128` or whatever) sounds kind of enticing, but it's more involved and would involve writing unsafe code, so I didn't do that for now. 

## Future work

- Expand the API to include shortcuts like `inverse_mul` and `inverse_transform` for efficiency reasons.
- Include more convenience constructors and methods (e.g. `from_xy`, `from_xyz`).
- Refactor `bevy_math::bounding` to use the isometry types.
- Add conversions to/from isometries for `Transform`/`GlobalTransform` in `bevy_transform`.

Jondolf reviewed (commented):
Thanks, I've been wanting isometry types for a while already! Left some review comments.

Some other things I would like to be added:

- `inverse_mul` to compute `iso1.inverse() * iso2` more efficiently. This avoids an extra rotation.
- `inverse_transform_point` to compute `iso.inverse().transform_point(point)` more efficiently. This avoids an extra rotation.
- Probably `Mul` implementations for vectors (treated as points) and maybe directions (just rotated). `Transform` has `Mul<Vec3>` too, and it's much more terse than `transform_point` (although for isometries, perhaps a bit more implicit behavior-wise).
- `Isometry2d::from_xy` and `Isometry3d::from_xyz`
- A way to extract isometries from transforms
- A way to create transforms from isometries

These probably aren't blocking though, and could also be done in a follow-up.

Then there are other design considerations:

- Should we go with the `Isometry2d` or `Iso2` naming? Either is honestly fine to me though, and I don't want this to get stuck in endless bikeshedding, so we can probably leave this decision for later and just go with `Isometry2d/3d` for now.
- Where do we use these isometry types? We could already start using them for things like bounding volume computations. Actually changing these APIs can be left for a follow-up though.

NiseVoid commented on crates/bevy_math/src/isometry.rs:
```diff
@@ -0,0 +1,187 @@
+use std::ops::Mul;
+
+/// An isometry in two dimensions.
+#[derive(Copy, Clone, Debug, PartialEq)]
```
Should we have a `Default`, `Serialize` (behind the feature flag), and `Reflect` (also behind a feature flag)?

NiseVoid reviewed (commented):
Looks good overall, I agree with some of Jondolf's comments. It would probably be good to have a `Mul<VecX>` and `Mul<DirX>`.

Could we get some tests for the mul, transform and inverse methods?

mweatherley commented on crates/bevy_math/src/isometry.rs:
```diff
@@ -0,0 +1,187 @@
+
+    /// Create a two-dimensional isometry from a rotation and a translation.
+    #[inline]
+    pub fn from_rotation_translation(rotation: Rot2, translation: Vec2) -> Self {
```
Yeah, I agree. The only reason it was this way was to match the `Affine` types. I think I agree with not trying to adhere to that, so I'll basically just undo a bunch of changes that I made to make them match and go with things that adhere more closely to our own use.

mweatherley commented:
I think the changes should be pretty clear from the commit history, but here's what I did:
- Deviated from `Affine` conventions to rename to just `IsometryNd::new` and `transform_point`, reordering parameters as well.
- Added all of the standard `#[derive]` stuff to each new struct.
- Added `Mul<VecN>` (essentially uses `transform_point`, but the return type always matches the input type) and `Mul<DirN>` (ignores translation component) impls. 
- Added feature-conditioned impls for `AbsDiffEq`, `RelativeEq`, and `UlpsEq`, which all just look at the components.
- Added some tests.

Jondolf reviewed (approved):
LGTM as a solid starting point! We can add more methods and adapt some APIs to use these isometries in follow-ups.

alice-i-cecile commented:
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1656 if you'd like to help out.
